### ISSUE TITLE
Service relink doesn't trigger HAProxy config reload, leading to 503

### ISSUE THREAD

**Issue by yunspace** [tutumcloud/haproxy#119]
I've raised this in the forums https://support.tutum.co/support/discussions/topics/5000067035 and spoke to @tifayuki on intercom. But thought I'd raise an official ticket for record keeping and so others are aware of the issue also.

We have automatic blue green deployment setup as per http://blog.tutum.co/2015/06/08/blue-green-deployment-using-containers/ 

However we are seeing the behaviour that after a `Services` is re-linked (from say blue to green), there is no activity in the `HAProxy` logs which indicate the configs have not reloaded and the service is still pointing to the old version (blue). This is against the behaviour described in http://blog.tutum.co/2015/04/07/presenting-tutum-stream-api/ where when a `Service` is changed, the `HAProxy` should restart. 

This behaviour happens in both 0.1 and 0.2 releases of the `tutum/haproxy`. And applies to relinking a service via either Tutum API or editing the Stack file. 

The side effect is that because we shutdown the old service upon a successful relink returned by the API, the `HAProxy` now redirects traffic to a dead blue service which gives `503` error. 

This is quite a critical error and effects our live user facing applications. Our current work around is to force-restart the `HAProxy` every time we deploy but it's not ideal. Please apply a fix or alternatively expose a end point in `HAProxy` that can be used to kick off the config reload without restarting `HAProxy` containers.

**Comment by nclaw** [comment]
We have tried a workaround suggested by one of users on tutum slack group - instead of relinking HAproxy from green to blue and vice versa we should constantly have links from HAproxy to both green and blue services and scale down one of the services (ex: blue) after another one has been redeployed (ex: green).

However this method did not work as expected and produced following results:
- When both blue and green services are up - HAproxy runs with no issues;
- When either of the two servcies is in state "not running" or "stopped" HAproxy produces following error: "ERROR:root:u'SERVICE_NAME_GREEN'"

**Comment by tifayuki** [comment]
The issue that prevents `tutum/haproxy` from responding to new event is fixed: https://github.com/tutumcloud/haproxy/commit/72b46679b697dbddaf036257cc76d6372bb3f4c1

If you set the same VIRTUAL_HOST on GRREE/BLUE service, the containers will be merged in the same backend together. Then, you can gradually scale down one of them to zero. After the change is stable, you can safely remove the link.

Also, I will start versioning the release, so that we know which bug happens to which code.

**Comment by yunspace** [comment]
Thanks @tifayuki. We've managed to resolve this by terminating the HAProxy and starting a new service with exact same settings, everything now works as expected. Events were fired upon link update and configs were reloaded. However it still didn't work in the old HAProxy. 

Very strange behaviour, I speculating perhaps it's because this particular HAProxy was created before the Stream API was introduced in June and therefore doesn't receive Stream events? I will observe further and re-open this ticket if it occurs again.

**Comment by yunspace** [comment]
This is issue is still happening for us. Using a completely new HAProxy using tutum/haproxy:staging and some times we would see the correct behaviour, where each relink event would log something like:

```
router-qa-1 | 2015-11-30T06:20:11.073590198Z INFO:haproxy:Tutum event: service 0099ac99-b512-4e0a-89bf-6e8acd03a2e6 is unlinked from HAProxy service 5bcf5528-cba3-471c-ab33-518af71f4104 is linked to HAProxy
router-qa-1 | 2015-11-30T06:20:16.653186966Z INFO:haproxy:Current links: blueprint-service-qa-blue(6c3703bb-d94c-4810-aedb-fc169666366c), file-service-qa-blue(d2a70235-61e6-4e3f-8089-d237c2fb3592), identity-service-qa-green(10f31d53-9795-4475-87f6-f64e2fa71867), indexing-service-qa(630d0774-0ade-45c2-a737-ba7f8a9b11df), notification-service-qa-blue(84cecfb9-f4f8-4b7c-9cc6-e05758742e41), opengraph-service-qa(5e83bc6d-e444-4b81-a654-83e75c8c7c54), search-service-qa-blue(ff50be61-5779-459b-87d8-57883fca0af1), stream-service-qa-blue(6e68f236-e817-4b4f-a1cf-7389e1a4f49c), tasks-service-qa-green(903e741f-d5e3-40d0-b2e3-2f03f579005a), trunk-io-qa-blue(5bcf5528-cba3-471c-ab33-518af71f4104), webhook-service-qa-green(d6d00648-1e3b-4e50-9dd3-9d6b81a456fd)
router-qa-1 | 2015-11-30T06:20:16.781092972Z INFO:haproxy:HAProxy configuration is updated:
```

However every now and then, HAProxy would just stop receiving events. This problem would occur multiple times in a day. 

   * Task service from blue to green at 9:16AM UTC, 30th of November (8PM Sydney time):
```
  I, [2015-11-30T09:16:32.803962 #4363]  INFO -- Tutum Deployment: Deployment status: Success)
  I, [2015-11-30T09:16:36.724428 #4363]  INFO -- Tutum Deployment: tasks-service-dev.trunk-dev-green.trunkbot.svc.tutum.io running healthy
  I, [2015-11-30T09:16:38.753945 #4363]  INFO -- Tutum Deployment: switching router router-dev to use tasks-service-dev.trunk-dev-green.trunkbot.svc.tutum.io
  I, [2015-11-30T09:16:51.756084 #4363]  INFO -- Tutum Deployment: Action Success
  I, [2015-11-30T09:16:51.756639 #4363]  INFO -- Tutum Deployment: router switched tasks-service-dev.trunk-dev-green.trunkbot.svc.tutum.io, shutting down tasks-service-dev.trunk-dev-blue.trunkbot.svc.tutum.io
```

  * Trunk service from blue to green at 9:40PM UTC, 30th of November (8AM Sydney time, 1st December):
```
I, [2015-11-30T21:40:27.220736 #3577]  INFO -- Tutum Deployment: trunk-io-dev.trunk-dev-green.trunkbot.svc.tutum.io running healthy
I, [2015-11-30T21:40:29.765561 #3577]  INFO -- Tutum Deployment: switching router router-dev to use trunk-io-dev.trunk-dev-green.trunkbot.svc.tutum.io
I, [2015-11-30T21:40:40.561457 #3577]  INFO -- Tutum Deployment: Action Success
I, [2015-11-30T21:40:40.561559 #3577]  INFO -- Tutum Deployment: router switched trunk-io-dev.trunk-dev-green.trunkbot.svc.tutum.io, shutting down trunk-io-dev.trunk-dev-blue.trunkbot.svc.tutum.io
```

  * Tutum stack shows the links have been updated:
![screen shot 2015-12-01 at 10 33 24 am](https://cloud.githubusercontent.com/assets/218128/11488438/9e981f88-981a-11e5-8402-ae615c73abea.png)

  * But the logs hasn't shown any updates since 5:14AM UTC, 30 of November (4:14PM Sydney time). Both services are still pointing to BLUE for the 17 hours and didn't register the 2 events that happened in between. Is there anyway to check if the events were fired at the above times?

```
router-dev-1 | 2015-11-30T05:14:52.230848497Z backend SERVICE_TASKS_SERVICE_DEV_BLUE
router-dev-1 | 2015-11-30T05:14:52.230853531Z   redirect scheme https code 301 if !{ ssl_fc }
router-dev-1 | 2015-11-30T05:14:52.230858540Z   server TASKS_SERVICE_DEV_BLUE_1 10.7.0.1:80 check
router-dev-1 | 2015-11-30T05:14:52.230865912Z backend SERVICE_TRUNK_IO_DEV_BLUE
router-dev-1 | 2015-11-30T05:14:52.230870832Z   redirect scheme https code 301 if !{ ssl_fc }
router-dev-1 | 2015-11-30T05:14:52.230875776Z   server TRUNK_IO_DEV_BLUE_1 10.7.0.17:80 check
router-dev-1 | 2015-11-30T05:14:52.230880561Z   server TRUNK_IO_DEV_BLUE_2 10.7.0.61:80 check

router-dev-2 | 2015-11-30T05:14:51.731042991Z backend SERVICE_TASKS_SERVICE_DEV_BLUE
router-dev-2 | 2015-11-30T05:14:51.731047774Z   redirect scheme https code 301 if !{ ssl_fc }
router-dev-2 | 2015-11-30T05:14:51.731052650Z   server TASKS_SERVICE_DEV_BLUE_1 10.7.0.1:80 check
router-dev-2 | 2015-11-30T05:14:51.731124052Z backend SERVICE_TRUNK_IO_DEV_BLUE
router-dev-2 | 2015-11-30T05:14:51.731132786Z   redirect scheme https code 301 if !{ ssl_fc }
router-dev-2 | 2015-11-30T05:14:51.731138651Z   server TRUNK_IO_DEV_BLUE_1 10.7.0.17:80 check
router-dev-2 | 2015-11-30T05:14:51.731143743Z   server TRUNK_IO_DEV_BLUE_2 10.7.0.61:80 check
```

This happens so frequently that I'm confident if we organise a call, I can reproduce this simply by repeating a deployment a couple of times. Please investigate. Our username is `trunkbot`. 

This really is a show stopper for us, we will be talking to Docker folks about tutum's GA plans this week. It would be good if a engineer could be present also. Will update call details later.

**Comment by yunspace** [comment]
I have a theory. The periodic issues we are seeing could potentially be due to a race condition between the StreamAPI event, `main.py` and HAProxy service.

Basically:

1. An update action completes successfully for switching service link A from BLUE to GREEN, an event is fired off.
2. The HAProxy service is still updating, therefore link A is still BLUE
3. `main.py` receives the event, instead of relying on the `Action.body.linked_to_service`, it pulls out  `Service.linked_to_service` which at this point could still have the old BLUE link.
4. `main.py` compares `service_endpoints` with `cls_linked_services`, does nothing.
5. HAProxy service finishes updating itself and shows link A as GREEN. But the config isn't reloaded.

I believe the fix would be to replace:

https://github.com/tutumcloud/haproxy/blob/master/haproxy/main.py#L40
https://github.com/tutumcloud/haproxy/blob/master/haproxy/main.py#L41

with: 

        event_action = Haproxy.fetch_tutum_obj(Haproxy.fetch_tutum_obj(event.get("resource_uri"))
        action_endpoints = [srv.get("to_service") for srv in event_action.body.get("linked_to_service")]

So basically trusting that if the event is "Successful" then the `Action` will be completed and rely on `linked_to_service` from the `Action` instead of prematurely pull it from the service.

**Comment by cggaurav** [comment]
@yunspace I second this, we have a proxy for our api and web app and on scaling down/up - our HAproxy doesn't reload. We have 'dead' api containers in our stats page. ![](https://d1zjcuqflbd5k.cloudfront.net/files/acc_279689/1dzFK?response-content-disposition=inline;%20filename=Screen%20Shot%202015-12-08%20at%2013.28.21.png&Expires=1449578533&Signature=RCvjhg~NYuc3WM5G9~~0wbFwuoU~0~120Aazd4~91lsgvkQKoo5zoxoYZiaDtv9VEt~TIEwFJ3V782198vjAPl6Pyk0foZZPBVu0K9EtnYq4pAo9cJch80Sayoi4DtELsh7LN~pNnt~RJWE~uiBeQtNBfHPmxm9PNmiIV-eHQ0Y_&Key-Pair-Id=APKAJTEIOJM3LSMN33SA)

**Comment by yunspace** [comment]
@cggaurav the code responsible for the reload upon scaling is line 30-36 in `main.py`. The approach we took we is add/remove the linke, which touches on lines 38-53 in `main.py`. 

Have a look into the logs and see if you get any "Tutum event" entries?

**Comment by yunspace** [comment]
I made the changes I described here: https://github.com/Trunkplatform/haproxy/blob/events/haproxy/main.py#L40 Will test this out in our dev environment and let you know how it goes.

**Comment by stephenkeep** [comment]
@yunspace did the fix work we are having a similar problem with linking new containers.

**Comment by yunspace** [comment]
Hi @stephenkeep unfortunately it didn't. I deployed my code and also took extra measures to write a simple event listener https://github.com/Trunkplatform/tutum-event-client just for this issue. 

Basically it wasn't a matter of race condition between the `event` and the `service`, when Haproxy fails to reload, it is because there was no event published at all. I can prove it now since my `tutum-event-client` listens to everything and both my listener and `HAProxy` logged nothing. 

This is quite reproducible, listen to tutum events, and switch links using either the API or GUI, every re-link should fire a event that has `action=ServiceUpdate` and `parent=Haproxy uuid`. Keep doing it for a while and eventually no event will fire. 

When I spoke to @borjaburgos previously he mentioned `Haproxy` is an opensource effort and isn't part of the core service and therefore not under SLA, which is understandable. However if events are missing on an ad-hoc basis then I would assume it should be included in the SLA since the `StreamApi` is part of tutum's core services. 

Anyway I will revert my unnecessary `race condition` changes. Then apply a work around to fire a custom event from my zero-downtime deployment script every time I re-link. And have HAProxy listen to my custom event to reload itself.

**Comment by yunspace** [comment]
So here are my logs which makes me certain the event is not being fired, and it's not the HAProxy code:

Deployment and service relink at 03:58 UTC (14:58 Sydney time): My event Listener picks up `Service Update` action event:

```
Dec 11 14:58:43 logspout-dev-1 tutum-event-client-1.devop-stack.9d4a4368:  [[11/Dec/2015 03:58:43.356]] [LOG]   {"state":"Success","parents":["/api/v1/service/193306e2-276d-4d6c-968c-c36f59248d0f/","/api/v1/stack/1ea631da-e9a1-4f0b-9da3-659e592d66fa/"],"action":"update","type":"action","resource_uri":"/api/v1/action/319c23a3-4665-414f-b530-7ec305c9f12d/","uuid":"4cc45af8-c58e-4593-9441-cefbb6d65d1d","date":"2015-12-11T03:58:43+00:00"}
```

And `HAProxy` also reloads itself upon receiving the event:

```
router-dev-2 | 2015-12-11T03:58:50.685044986Z   redirect scheme https code 301 if !{ ssl_fc }
router-dev-2 | 2015-12-11T03:58:50.685061250Z   server TRUNK_IO_DEV_GREEN_1 10.7.0.74:80 check
router-dev-2 | 2015-12-11T03:58:50.685069869Z   server TRUNK_IO_DEV_GREEN_2 10.7.0.75:80 check
router-dev-2 | 2015-12-11T03:58:50.685079522Z backend SERVICE_WEBHOOK_SERVICE_DEV_BLUE
router-dev-2 | 2015-12-11T03:58:50.685089424Z   redirect scheme https code 301 if !{ ssl_fc }
router-dev-2 | 2015-12-11T03:58:50.685099078Z   server WEBHOOK_SERVICE_DEV_BLUE_1 10.7.0.38:8080 check
router-dev-2 | 2015-12-11T03:58:50.685108902Z   server WEBHOOK_SERVICE_DEV_BLUE_2 10.7.0.68:8080 check
router-dev-2 | 2015-12-11T03:58:50.685118734Z INFO:haproxy:Reloading HAProxy
router-dev-2 | 2015-12-11T03:58:50.685128312Z INFO:haproxy:HAProxy has been reloaded(PID: 22)
router-dev-2 | 2015-12-11T03:58:50.685138360Z INFO:haproxy:===========END===========
router-dev-1 | 2015-12-11T03:58:51.297535393Z backend SERVICE_TRUNK_IO_DEV_GREEN
router-dev-1 | 2015-12-11T03:58:51.297544399Z   redirect scheme https code 301 if !{ ssl_fc }
router-dev-1 | 2015-12-11T03:58:51.297553183Z   server TRUNK_IO_DEV_GREEN_1 10.7.0.74:80 check
router-dev-1 | 2015-12-11T03:58:51.297562250Z   server TRUNK_IO_DEV_GREEN_2 10.7.0.75:80 check
router-dev-1 | 2015-12-11T03:58:51.297570422Z backend SERVICE_WEBHOOK_SERVICE_DEV_BLUE
router-dev-1 | 2015-12-11T03:58:51.297578572Z   redirect scheme https code 301 if !{ ssl_fc }
router-dev-1 | 2015-12-11T03:58:51.297586912Z   server WEBHOOK_SERVICE_DEV_BLUE_1 10.7.0.38:8080 check
router-dev-1 | 2015-12-11T03:58:51.297595552Z   server WEBHOOK_SERVICE_DEV_BLUE_2 10.7.0.68:8080 check
router-dev-1 | 2015-12-11T03:58:51.297664526Z INFO:haproxy:Reloading HAProxy
router-dev-1 | 2015-12-11T03:58:51.328326344Z INFO:haproxy:HAProxy has been reloaded(PID: 18)
router-dev-1 | 2015-12-11T03:58:51.328399935Z INFO:haproxy:===========END===========
router-dev-2 | 2015-12-11T03:58:54.087731647Z INFO:haproxy:HAProxy(PID:18) has been terminated
router-dev-1 | 2015-12-11T03:58:54.266172826Z INFO:haproxy:HAProxy(PID:14) has been terminated
router-dev-2 | 2015-12-11T03:59:07.491820177Z INFO:haproxy:HAProxy(PID:20) has been terminated
router-dev-1 | 2015-12-11T03:59:36.640379645Z INFO:haproxy:HAProxy(PID:16) has been terminated
```

Deployment and service relink at 05:03 UTC (16:03 Sydney time), resulted in not event registered in my listener nor HAProxy: 

```
I, [2015-12-11T05:02:58.197184 #13556]  INFO -- Tutum Deployment: trunk-io-dev.trunk-dev-blue.trunkbot.svc.tutum.io running healthy
I, [2015-12-11T05:03:01.245572 #13556]  INFO -- Tutum Deployment: switching router router-dev to use trunk-io-dev.trunk-dev-blue.trunkbot.svc.tutum.io
I, [2015-12-11T05:03:16.454869 #13556]  INFO -- Tutum Deployment: Action Success
I, [2015-12-11T05:03:16.455632 #13556]  INFO -- Tutum Deployment: router switched trunk-io-dev.trunk-dev-blue.trunkbot.svc.tutum.io, shutting down trunk-io-dev.trunk-dev-green.trunkbot.svc.tutum.io
I, [2015-12-11T05:03:34.035499 #13556]  INFO -- Tutum Deployment: Action Success
```

**Comment by tifayuki** [comment]
@yunspace 
We made some changes on the stream API, which should fix the missing event issue. Can you try if it works?

**Comment by yunspace** [comment]
@tifayuki thanks for that. I'm a bit paranoid of this issue now since it's a pain in the backside when users start getting 503s in production. So I've actually added a fail safe manual reload endpoint:

    http://<haproxy.hostname>:5000/main/reload

The code is here: https://github.com/Trunkplatform/haproxy/blob/api_reload/haproxy/main.py#L83

Just like `reload.sh` it won't reload the config if there are no changes. But this endpoint will be useful when `docker exec` or `tutum exec` are not available to run the `reload.sh`, and can be called easily by Continuous Integration/Delivery scripts. 

I'm planning to add some authentication to this (perhaps using the static page credentials) and then raise a PR for you to consider.

**Comment by thekiwi** [comment]
@yunspace and @tifayuki - we've been struggling with the same issue for several months now, regular 503s when the services are clearly running post-deploy, and the haproxy service not detecting the new container or reloading configuration. 

Did you get any further with solving the issue or are you still resorting to the manual-reload script?

**Comment by yunspace** [comment]
hi @thekiwi I'm not using `tutum` on a regular basis anymore as I have moved roles. However I do believe @tifayuki has made a fix to the `tutum stream api` and has addressed the issue I'm having with events not triggering service reloads. 

When a 503 occurs I suggest you have a look at the haproxy stats page and verify all of your services are there and coloured green. 

I'll close this for now. You may need to work with the tutum support folks further on the specific of your issue.
